File: javascript150/15.ts
function threeSum(nums: number[]): number[][] {
    nums = nums.sort((a, b) => a-b)
    let answer: number[][] = []

    for (let a = 0; a < nums.length-1; a++) {
        if (a > 0 && nums[a] === nums[a - 1]) continue;
        let left = a+1, right = nums.length-1

        while (left < right) {
            let total = nums[a] + nums[left] + nums[right]

            if (total < 0) {
                left++
            } else if (total > 0) {
                right--
            } else {
                let ans = [nums[a], nums[left], nums[right]].sort((a,b)=>a-b)
                answer.push(ans)

                left++
                while (left < right && nums[left] == ans[1]) {
                    left++
                }

                right--
                while (left < right && nums[right] == ans[2]) {
                    right--
                }
            }
        }
    }
    return answer;
};

console.log(threeSum([-1,0,1,2,-1,-4]))

File: RandomPractice/15.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        nums.sort()

        a = 0
        answer = []
        
        while a < len(nums) - 2:
            b = a+1
            c = len(nums)-1

            while b < c:
                total = nums[a] + nums[b] + nums[c]
                
                if total == 0:
                    if [nums[a], nums[b], nums[c]] not in answer:
                        answer.append([nums[a], nums[b], nums[c]])
                    b += 1

                elif total < 0:
                    b += 1
                    while b < len(nums)-1 and nums[b-1] == nums[b]:
                        b += 1

                else:
                    c -= 1
                    while c >= 0 and nums[c] == nums[c+1]:
                        c -= 1

            a += 1
            while a < len(nums)-1 and nums[a-1] == nums[a]:
                a += 1
        
        return answer
    
sol = Solution()
# print(sol.threeSum([0,0,0]))
print(sol.threeSum([1,2,0,1,0,0,0,0]))
# print(sol.threeSum([-1,0,1,2,-1,-4]))